iroh-net/src/relay/codec.rs

```diff
@@ -264,6 +264,57 @@ impl Frame {
         }
     }
 
+    fn from_ws_vec(vec: Vec<u8>) -> anyhow::Result<Self> {
+        if vec.is_empty() {
+            bail!("error parsing relay::codec::Frame: too few bytes (0)");
+        }
+        let bytes = Bytes::from(vec);
+        let typ = FrameType::from(bytes[0]);
+        let frame = Self::from_bytes(typ, bytes.slice(1..))?;
```

Comment:
could skip allocating the first byte entirely

```rust
let typ = vec[0];
let bytes = Bytes::from(&vec[1..]);
let frame = Self::from_bytes(typ, bytes.)?;
```

Reply:
Are we sure that `Bytes::from(&vec[1..])` doesn't copy the vec contents?
I don't see how it can own the underlying bytes, if what's passed in is some slice with a lifetime. :eyes:

Reply:
from vec already copies unfortunately

Reply:
I'm pretty sure it doesn't copy...
This is the implementation: https://docs.rs/bytes/latest/src/bytes/bytes.rs.html#830-862
Here's a playground where I'm comparing/printing the pointers: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a2b54d45c3e50b4e96e77ecfaf73a746

Reply:
ah right `Bytes` doesn't allocate, but `BytesMut` does

Reply:
(I think)

Reply:
Yep! (from a short peek at [the source](https://docs.rs/bytes/latest/src/bytes/bytes_mut.rs.html#1615-1648))